packages/@react-stately/data/src/useAsyncList.ts

```diff
@@ -200,12 +202,15 @@ function reducer<T, C>(data: AsyncListState<T, C>, action: Action<T, C>): AsyncL
     case 'loadingMore':
       switch (action.type) {
         case 'success':
+          selectedKeys = (data.selectedKeys === 'all' || action.selectedKeys === 'all')
+            ? 'all'
+            : [...data.selectedKeys, ...(action.selectedKeys ?? [])];
           // Append the new items
           return {
             ...data,
             state: 'idle',
             items: [...data.items, ...action.items],
-            selectedKeys: new Set([...data.selectedKeys, ...(action.selectedKeys ?? [])]),
+            selectedKeys: selectedKeys === 'all' ? 'all' : new Set(selectedKeys),
```

Comment:
Small nit: you could move the `new Set` up to the selectedKeys line and get rid of the ternary on 213
i.e.
```
          selectedKeys = (data.selectedKeys === 'all' || action.selectedKeys === 'all')
            ? 'all'
            : new Set([...data.selectedKeys, ...(action.selectedKeys ?? [])]);
...
        selectedKeys: selectedKeys,
```